### PR TITLE
adds default alt values to inline images and the image block

### LINKED ISSUE #1520: Images alt attribute in an editing context
Splitting this out from #1309 

One interesting thing we've recently discussed in the accessibility team, during the review of the new media widgets, is the different purpose the alt text has in the front-end and in an editing context (ie. when editing a post).

As a general rule, we all know that (I'll simplify a bit):
- purely decorative images should have an empty alt text `alt=""`
- images that convey relevant information, should use a meaningful alt text

However, when in an editing context, things are slightly different. Will try to do my best to explain. While editing a post, users need to know what the image they've added is. 
When users add an alt text to the image because they want to communicate the image purpose in the front-end, that also serves the purpose to make the image recognizable in the editor.
However, when the image is purely decorative, users will set an empty alt text. That's perfectly correct for the front-end but it will also make the image not recognizable for screen reader users because images with an `alt=""` are just ignored.

Ideally, when in the editor, also images intended to use an empty `alt=""` should have _something_ to make them perceivable and announced by screen readers. Not sure about the best option, but maybe this is something worth considering as a nice improvement Gutenberg could bring in.

As a reference, the solution implemented for the image widget was to remove the empty alt attribute in the editor when there's no alt value, so at least the filename gets announced. Not sure that's the best available option and maybe we could consider an improved solution.
https://github.com/xwp/wp-core-media-widgets/issues/47
https://github.com/xwp/wp-core-media-widgets/pull/50

### PR DESCRIPTION
**Pull request by antpb**
Sets default alt values to fix #1520 
Previous conversation can be found here: https://github.com/WordPress/gutenberg/pull/10482

How to test:
create image block
add image with no alt value
inspect image to see that alt value is now set as a readable string "This image has an empty alt attribute, its file name is "FILENAME""

### CODE REVIEW

File: packages/editor/src/components/rich-text/core-tokens/image/index.js [hunk 1]
```diff
@@ -35,7 +35,11 @@ export const settings = {
 		);
 	},
 
-	save( { id, url, alt, width } ) {
+	save( { id, url, alt, width, filename } ) {
+		if ( ! alt ) {
+			alt = sprintf( __( 'This image has an empty alt attribute; its file name is "%s"' ), filename );
```
Comment (ocean90) [inline packages/editor/src/components/rich-text/core-tokens/image/index.js]: The issue #1520 is only about adding this text to images in editing context. The `save()` method is what gets saved into the database and is visible on the front end. So this change seems wrong?

Comment (antpb) [reply packages/editor/src/components/rich-text/core-tokens/image/index.js]: I can change this to only be the edit context. I set the alt if it is empty on the save method because from my understanding, screen readers will try and read the full image url if there is no alt attribute defined.

Comment (ocean90) [reply packages/editor/src/components/rich-text/core-tokens/image/index.js]: Related chat on Slack where this has been discussed: https://wordpress.slack.com/archives/C02SX62S6/p1540406745000100

tldr;: empty alt texts are fine, suggested change should only apply in editing context.

**Comment by antpb** [comment]
@ocean90 Thanks for the catch. :) I've updated the PR to only apply the default alt text in the editing context if no `alt` is defined. I also tested adding alt text to the inspector settings and it updates from the default to the new inspector alt value as expected.

If you have any recommendations on how to better optimize this approach, I'd really appreciate it.

File: packages/block-library/src/image/edit.js [hunk 2]
```diff
@@ -512,6 +541,45 @@ class ImageEdit extends Component {
 								}
 							}
 							/* eslint-enable no-lonely-if */
+							if ( ! alt ) {
+								return (
```
Comment (aduth) [inline packages/block-library/src/image/edit.js]: There's a ton of code duplication here. What's actually different? Can we refactor to limit the variation of this specific condition? i.e. if it's to return `imgNoAlt` instead of `img`, just change that in the existing return value:

```
{ alt ? img : imgNoAlt }
```

File: packages/block-library/src/image/edit.js [hunk 3]
```diff
@@ -58,7 +58,9 @@ const LINK_DESTINATION_CUSTOM = 'custom';
 const ALLOWED_MEDIA_TYPES = [ 'image' ];
 
 export const pickRelevantMediaFiles = ( image ) => {
-	return pick( image, [ 'alt', 'id', 'link', 'url', 'caption' ] );
+	return {
+		...pick( image, [ 'alt', 'id', 'link', 'url', 'caption' ] ),
```
Comment (aduth) [inline packages/block-library/src/image/edit.js]: Why spread.

**Comment by aduth** [comment]
You're going to have an easier time against merge conflicts if you rebase and keep to a single commit.

https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase

File: packages/block-library/src/image/edit.js (hunk 3 above)
Comment (antpb) [reply packages/block-library/src/image/edit.js]: I actually reverted this change prior to the nasty rebase attempt. The spread made sense in a prior iteration where I was changing the `alt` value. Thanks for calling out. :)

**Comment by antpb** [comment]
:( I made this PR really messy with that last rebase attempt. Sorry.  I've added a new, more final PR that should be easier to review/merge: https://github.com/WordPress/gutenberg/pull/11218
